java-checks/src/main/java/org/sonar/java/checks/serialization/SerializableObjectInSessionCheck.java

```diff
@@ -44,8 +46,44 @@
   protected void onMethodInvocationFound(MethodInvocationTree mit) {
     ExpressionTree argument = mit.arguments().get(1);
     Type type = argument.symbolType();
-    if (!type.isPrimitive() && !type.isSubtypeOf("java.io.Serializable")) {
-      reportIssue(argument, "Make \"" + type + "\" serializable or don't store it in the session.");
+    if (!isSerializable(type)) {
+      String andParameters = isParametrized(type) ? " and its parameters" : "";
+      reportIssue(argument, "Make \"" + type + "\"" + andParameters + " serializable or don't store it in the session.");
     }
   }
+
+  private static boolean isSerializable(Type type) {
+    if (type.isPrimitive()) {
+      return true;
+    }
+    if (isSerializableArray(type)) {
+      return true;
+    }
+    if (isParametrized(type)) {
+      return isSerializableParametrized((ParametrizedTypeJavaType) type);
+    }
+    return type.isSubtypeOf("java.io.Serializable");
+  }
+
+  private static boolean isSerializableArray(Type type) {
+    if (type instanceof ArrayJavaType) {
+      ArrayJavaType arrayJavaType = (ArrayJavaType) type;
+      return isSerializable(arrayJavaType.elementType());
+    }
+    return false;
+  }
+
+  private static boolean isParametrized(Type type) {
+    return type instanceof ParametrizedTypeJavaType;
+  }
+
+  private static boolean isSerializableParametrized(ParametrizedTypeJavaType type) {
+    // note: this is assuming that custom implementors of Collection
+    // have the good sense to make it serializable just like all implementations in the JDK
+    //
+    // note: type.substitution(t) should never be null
+    //noinspection ConstantConditions
```

Comment:
No need for the IDE-specific comment. we prefer to deal with the warnings manually.